taichi/gui/gui.h

```diff
@@ -777,6 +779,18 @@ class GUI : public GUIBase {
     }
     last_frame_time = taichi::Time::get_time();
     redraw();
+    // Some old examples / users don't even provide a `break` statement for us
+    // to terminate loop. So we have to terminate the program with RuntimeError
+    // if ti.GUI.EXIT event is not processed. Pretty like SIGTERM, you can hook
+    // it, but you have to terminate after your handler is done.
+    if (should_close) {
+      if (++should_close >
+          5) {  // if the event is not processed in 5 frames, raise RuntimeError
+        throw std::string(
```

Comment:
I haven't followed up on the Jupyter issue, so my comment here might not cover the whole picture. Is throwing an exception the best way to solve whatever Jupyter' issue is.? Below are what I know so far:

1. Please don't use exception in C++, it has caused lots of headaches. See #1059 , #655
2. According to the comment, I believe it's the examples that need to be fixed...This logic doesn't seem generic enough to be placed at the GUI level. Based on your understanding, is it possible to fix the examples for Jupyter users?

Reply:
It's not the examples need to be fixed, it's our users' legacy codes need to be fixed.
They may write:
```
while True:
  gui.show()
```
I want them to write `while gui.running`, but I can't, since they may not know this change.
We have to throw an exception in `gui.show()` to break the infinite loop, no other option, sadly.
BTW FYI `std::exception` here is not a C++ exception, it's a `RuntimeError` in Python exception, according to pybind11.

Reply:
I won't change my opinion that "if closed 5 times, quit the GUI" is an application level logic and doesn't belong to GUI. But I see your point that we need this to make the old code work. I'm fine with having this for Games 201, but I think eventually, we should aim for:

* make `running` read-only
* mark `running` to `false` if we've received the close event
* update the doc so that users know they should use `while gui.running:` if they want the program to be interactive

Hopefully this makes the code cleaner?

> FYI std::exception here is not a C++ exception, it's a RuntimeError in Python exception, according to pybind11.

Ah i see.

Reply:
t4po, will iapr.

> update the doc so that users know they should use while gui.running: if they want the program to be interactive

This is already done in both doc and error-message.